zig/download.sh

```diff
@@ -29,87 +25,75 @@ fi
-if command -v wget > /dev/null; then
-    # -4 forces `wget` to connect to ipv4 addresses, as ipv6 fails to resolve on certain distros.
-    # Only A records (for ipv4) are used in DNS:
-    ipv4="-4"
-    # But Alpine doesn't support this argument
-    if [ -f /etc/alpine-release ]; then
-    ipv4=""
-    fi
-    # shellcheck disable=SC2086 # We control ipv4 and it'll always either be empty or -4
-    ZIG_URL=$(wget $ipv4 --quiet -O - https://ziglang.org/download/index.json | grep -F "$ZIG_TARGET" | grep -F "$ZIG_RELEASE" | head -1 | awk '{print $2}' | sed 's/[",]//g')
-else
-    ZIG_URL=$(curl --silent https://ziglang.org/download/index.json | grep -F "$ZIG_TARGET" | grep -F "$ZIG_RELEASE" | head -1 | awk '{print $2}' | sed 's/[",]//g')
-fi
-
-# Ensure that the release is actually hosted on the ziglang.org website:
-if [ -z "$ZIG_URL" ]; then
-    echo "Release not found on ziglang.org"
-    exit 1
-fi
+ZIG_URL="https://ziglang.org/download/${ZIG_RELEASE}/zig-${ZIG_ARCH}-${ZIG_OS}-${ZIG_RELEASE}${ZIG_EXTENSION}"
+ZIG_CHECKSUM_EXPECTED=$(echo "$ZIG_CHECKSUMS" | grep -F "$ZIG_URL" | cut -d ' ' -f 2)
 
 # Work out the filename from the URL, as well as the directory without the ".tar.xz" file extension:
 ZIG_ARCHIVE=$(basename "$ZIG_URL")
-
-case "$ZIG_ARCHIVE" in
-    *".tar.xz")
-        ZIG_ARCHIVE_EXT=".tar.xz"
-        ;;
-    *".zip")
-        ZIG_ARCHIVE_EXT=".zip"
-        ;;
-    *)
-        echo "Unknown archive extension"
-        exit 1
-        ;;
-esac
-
-ZIG_DIRECTORY=$(basename "$ZIG_ARCHIVE" "$ZIG_ARCHIVE_EXT")
+ZIG_DIRECTORY=$(basename "$ZIG_ARCHIVE" "$ZIG_EXTENSION")
 
 # Download, making sure we download to the same output document, without wget adding "-1" etc. if the file was previously partially downloaded:
-echo "Downloading $ZIG_URL..."
-if command -v wget > /dev/null; then
+if command -v curl > /dev/null; then
+    curl --silent --output "$ZIG_ARCHIVE" "$ZIG_URL"
+elif command -v wget > /dev/null; then
     # -4 forces `wget` to connect to ipv4 addresses, as ipv6 fails to resolve on certain distros.
     # Only A records (for ipv4) are used in DNS:
     ipv4="-4"
     # But Alpine doesn't support this argument
     if [ -f /etc/alpine-release ]; then
-    ipv4=""
+        ipv4=""
     fi
+
     # shellcheck disable=SC2086 # We control ipv4 and it'll always either be empty or -4
     wget $ipv4 --quiet --output-document="$ZIG_ARCHIVE" "$ZIG_URL"
 else
-    curl --silent --output "$ZIG_ARCHIVE" "$ZIG_URL"
+    echo "Neither curl nor wget available."
+    exit 1
+fi
+
+# Verify the checksum.
+ZIG_CHECKSUM_ACTUAL=""
+if command -v sha256sum > /dev/null; then
+    ZIG_CHECKSUM_ACTUAL=$(sha256sum "$ZIG_ARCHIVE" | cut -d ' ' -f 1)
+elif command -v shasum > /dev/null; then
+    ZIG_CHECKSUM_ACTUAL=$(shasum -a 256 "$ZIG_ARCHIVE" | cut -d ' ' -f 1)
+else
+    echo "Neither sha256sum nor shasum available."
+    exit 1
+fi
```

Comment:
This is lovely! Now, what if we don't `rm "$ZIG_ARCHIVE"`, and _skip_ the download step if we already have an archive locally with the required checksum? Then, we could finally add a cache to our GitHub actions. 

I don't think caching is important for performance, downloading zig is pretty fast, but I feel bad about using public Zig infra in an inefficient way!